src/materials/LinearSpring.C

```diff
@@ -0,0 +1,247 @@
+    _rotations(declareProperty<RealVectorValue>("rotations")),
+    _kx(coupledValue("kx")),
+    _ky(coupledValue("ky")),
+    _kz(coupledValue("kz")),
+    _krx(coupledValue("krx")),
+    _kry(coupledValue("kry")),
+    _krz(coupledValue("krz")),
+    _spring_forces(declareProperty<RealVectorValue>("forces")),
+    _spring_moments(declareProperty<RealVectorValue>("moments")),
+    _kdd(declareProperty<RankTwoTensor>("displacement_stiffness_matrix")),
+    _krr(declareProperty<RankTwoTensor>("rotation_stiffness_matrix")),
+    _total_global_to_local_rotation(
+        declareProperty<RankTwoTensor>("total_global_to_local_rotation"))
+{
+  // Checking for consistency between length of the provided displacements and rotations vector
+  if (_ndisp != _nrot)
+    mooseError("LinearSpring: The number of variables supplied in 'displacements' "
+               "and 'rotations' input parameters must be equal.");
+
+  // Fetch coupled variables and gradients (as stateful properties if necessary)
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    MooseVariable * disp_variable = getVar("displacements", i);
+    _disp_num[i] = disp_variable->number(); // Displacement variable numbers in MOOSE
+
+    MooseVariable * rot_variable = getVar("rotations", i);
+    _rot_num[i] = rot_variable->number(); // Rotation variable numbers in MOOSE
+  }
+}
+
+void
+LinearSpring::computeQpProperties()
+{
+  // Compute initial orientation and length of the spring in global coordinate system
+  // Fetch the two nodes of the link element
+  std::vector<Node *> node;
+  for (unsigned int i = 0; i < 2; ++i)
+    node.push_back(_current_elem->get_node(i));
+  RealGradient x_orientation;
+  for (unsigned int i = 0; i < _ndisp; ++i)
+    x_orientation(i) = (*node[1])(i) - (*node[0])(i);
+  // _original_length[_qp] = x_orientation.norm();
+  x_orientation /= x_orientation.norm(); // Normalizing with length to get orientation
+
+  // Get y orientation of the spring in global coordinate system
+  RealGradient y_orientation = getParam<RealGradient>("y_orientation");
+  // Real dot = x_orientation(0) * y_orientation(0) + x_orientation(1) * y_orientation(1) +
+  //            x_orientation(2) * y_orientation(2);
+  Real dot = x_orientation * y_orientation;
+
+  // Check if x and y orientations are perpendicular
+  if (abs(dot) > 1e-4)
+    mooseError("Error in LinearSpring: y_orientation should be perpendicular to "
+               "the axis of the beam.");
+
+  // Calculate z orientation in the global coordinate system as a cross product of the x and y
+  // orientations
+  RealGradient z_orientation = x_orientation.cross(y_orientation);
+  // z_orientation(0) = (x_orientation(1) * y_orientation(2) - x_orientation(2) * y_orientation(1));
+  // z_orientation(1) = (x_orientation(2) * y_orientation(0) - x_orientation(0) * y_orientation(2));
+  // z_orientation(2) = (x_orientation(0) * y_orientation(1) - x_orientation(1) * y_orientation(0));
```

Comment:
Please remove above three lines for the spring.

Reply:
For the beam scenario, where I am connecting the base of the lumped mass model to the basemat, I had to use many beams which all had different y orientations but the same z_orientation. So, I added a z_orientation option to the beam element recently (not yet merged). If you think there might be similar scenarios using spring elements, it might be good to have the user provide either y_orientation or z_orientation for a particular spring element.

Reply:
Actually, you are correct. I am not sure why I thought z_orientation was required. It should give the same answer as long as the rotated coordinate system is calculated using X x Y = Z. Also, I tried switching the z_orientation to y_orientation in the connections to the basemat (for the fixed base NPP analysis) and the results were identical.

Reply:
Ok, great!